presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java

```diff
@@ -0,0 +1,242 @@
+import javax.management.Notification;
+import javax.management.NotificationListener;
+import javax.management.ObjectName;
+import javax.management.openmbean.CompositeData;
+
+import java.lang.management.GarbageCollectorMXBean;
+import java.lang.management.ManagementFactory;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.function.Function;
+
+import static com.facebook.presto.util.StringTableUtils.getTableStrings;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableListMultimap.toImmutableListMultimap;
+import static java.util.Objects.requireNonNull;
+import static java.util.Map.Entry;
+import static java.util.AbstractMap.SimpleEntry;
+
+
+public class GcStatusMonitor
+{
+    private static final Logger log = Logger.get(GcStatusMonitor.class);
+    private static final String GC_NOTIFICATION_TYPE = "com.sun.management.gc.notification";
+    private final NotificationListener notificationListener = (notification, ignored) -> onNotification(notification);
+    private final SqlTaskManager sqlTaskManager;
+
+    @Inject
+    public GcStatusMonitor(
+            SqlTaskManager sqlTaskManager)
+    {
+        this.sqlTaskManager = requireNonNull(sqlTaskManager, "sqlTaskManager must not be null");
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        for (GarbageCollectorMXBean mbean : ManagementFactory.getGarbageCollectorMXBeans()) {
+            if (mbean.getName().equals("TestingMBeanServer")) {
+                continue;
+            }
+
+            ObjectName objectName = mbean.getObjectName();
+            try {
+                ManagementFactory.getPlatformMBeanServer().addNotificationListener(
+                        objectName,
+                        notificationListener,
+                        null,
+                        null);
+            }
+            catch (JMException e) {
+                throw new RuntimeException("Unable to add GC listener", e);
+            }
+        }
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        for (GarbageCollectorMXBean mbean : ManagementFactory.getGarbageCollectorMXBeans()) {
+            ObjectName objectName = mbean.getObjectName();
+            try {
+                ManagementFactory.getPlatformMBeanServer().removeNotificationListener(objectName, notificationListener);
+            }
+            catch (JMException ignored) {
+            }
+        }
+    }
+
+    private void onNotification(Notification notification)
+    {
+        if (GC_NOTIFICATION_TYPE.equals(notification.getType())) {
+            GarbageCollectionNotificationInfo info = new GarbageCollectionNotificationInfo((CompositeData) notification.getUserData());
+            if (info.isMajorGc()) {
+                onMajorGc();
+            }
+        }
+    }
+
+    private void onMajorGc()
+    {
+        try {
+            logActiveTasks();
```

Comment:
Think to think about: do we think it is fine to do the logging in the notification thread?

Reply:
Java has a service thread that calls this. The existing airlift GC status monitor is also logging from this and it doesn't seem to be causing issues.